### ISSUE TITLE
The performance of FusedBatchNormGradGrad is very low

### ISSUE THREAD

**Issue by mpjlu** [tensorflow/tensorflow#22080]
When I test the performance of GAN model (https://github.com/tensorflow/models/tree/master/research/gan), I find to use FusedBatchNorm can boost the performance by 20% to 30% (https://github.com/tensorflow/models/issues/5206).

However, when use gan_loss with gradient_penalty, there is no performance improvement with FusedBatchNorm.  I find the root cause is FusedBatchNormGradGrad is very low performance.

I propose to set fused_batch_norm to false on (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/gan/python/losses/python/losses_impl.py#L367) to use BatchNormGradGrad to replace FusedBatchNormGradGrad. There is about 20% performance improvement with this change. 
Change
`disc_interpolates = discriminator_fn(interpolates, generator_inputs)`
to 
`disc_interpolates = discriminator_fn(interpolates, generator_inputs, fused_batch_norm=false)`

**Comment by mpjlu** [comment]
@joel-shor , any comments for this, thanks.

**Comment by joel-shor** [comment]
This is very strange. I'd like to investigate. How did you benchmark the performance degradation? If it was code, would you mind pasting the code you used to benchmark?

**Comment by mpjlu** [comment]
Just run https://github.com/tensorflow/models/blob/master/research/gan/cifar/train.py with some CPU specific setting. 
I test on Intel CPU, no GPU.

**Comment by joel-shor** [comment]
Can you paste your sample output?

**Comment by mpjlu** [comment]
**Before change:**

./cifar/train.py  -dl /dataset/tensorflow/cifar10/ --num_inter_threads 2 --num_intra_threads 28 --batch_size 32
INFO:tensorflow:SyncReplicasV2: replicas_to_aggregate=9; total_num_replicas=10
INFO:tensorflow:SyncReplicasV2: replicas_to_aggregate=9; total_num_replicas=10
WARNING:tensorflow:update_ops in create_train_op does not contain all the  update_ops in GraphKeys.UPDATE_OPS
WARNING:tensorflow:update_ops in create_train_op does not contain all the  update_ops in GraphKeys.UPDATE_OPS
==========inter 2 intra: 28
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
2018-09-05 16:35:18.421355: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX512F
INFO:tensorflow:Restoring parameters from ./cifar-train-log/unconditional/model.ckpt-2200
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 2200 into ./cifar-train-log/unconditional/model.ckpt.

User settings:

   KMP_AFFINITY=granularity=fine,verbose,compact,1,0
   KMP_BLOCKTIME=4
   KMP_SETTINGS=1
   OMP_NUM_THREADS=28

INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2200
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2210 (1.506 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2220 (1.386 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2230 (1.368 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2240 (1.390 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2250 (1.401 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2260 (1.437 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2270 (1.462 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2280 (1.436 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 2290 (1.417 sec)
INFO:tensorflow:global_step/sec: 6.78513

**After Change:**
python ./cifar/train.py  -dl /dataset/tensorflow/cifar10/ --num_inter_threads 2 --num_intra_threads 28 --batch_size 32
INFO:tensorflow:SyncReplicasV2: replicas_to_aggregate=9; total_num_replicas=10
INFO:tensorflow:SyncReplicasV2: replicas_to_aggregate=9; total_num_replicas=10
WARNING:tensorflow:update_ops in create_train_op does not contain all the  update_ops in GraphKeys.UPDATE_OPS
WARNING:tensorflow:update_ops in create_train_op does not contain all the  update_ops in GraphKeys.UPDATE_OPS
==========inter 2 intra: 28
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
2018-09-05 16:51:42.265883: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX512F
INFO:tensorflow:Restoring parameters from ./cifar-train-log/unconditional/model.ckpt-500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 500 into ./cifar-train-log/unconditional/model.ckpt.

User settings:

   KMP_AFFINITY=granularity=fine,verbose,compact,1,0
   KMP_BLOCKTIME=4
   KMP_SETTINGS=1
   OMP_NUM_THREADS=28

INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 700 (1.292 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 710 (1.266 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 720 (1.193 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 730 (1.222 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 740 (1.248 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 750 (1.256 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 760 (1.235 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 770 (1.223 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 780 (1.244 sec)
INFO:tensorflow:Tensor("status_message:0", shape=(), dtype=string) = Starting train step: 790 (1.201 sec)
INFO:tensorflow:global_step/sec: 8.10213

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by DRACOyu** [comment]
WARNING:tensorflow:update_ops in create_train_op does not contain all the update_ops in GraphKeys.UPDATE_OPS
WARNING:tensorflow:update_ops in create_train_op does not contain all the update_ops in GraphKeys.UPDATE_OPS
Do you solve this problems? I have the same question,please give me some advice to solve this problem,Thanks.

**Comment by mpjlu** [comment]
I propose to set fused_batch_norm to false on (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/gan/python/losses/python/losses_impl.py#L367) to use BatchNormGradGrad to replace FusedBatchNormGradGrad. There is about 20% performance improvement with this change.
Change
disc_interpolates = discriminator_fn(interpolates, generator_inputs)
to
disc_interpolates = discriminator_fn(interpolates, generator_inputs, fused_batch_norm=false)

Here is a workaround for this problem.

**Comment by DRACOyu** [comment]
print(tf.get_collection(tf.GraphKeys.UPDATE_OPS)) is tf.operation.but print(tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)) is tf.variable,why,

**Comment by gadagashwini** [comment]
Hi @mpjlu, In Tensorflow v2.x, `Contrib` packages are deprecated. New upgraded TensorFlow-GAN (TF-GAN) is out now with lot of enhancements in performance and bug wise. Request you to try TF-GAN model from [here](https://github.com/tensorflow/gan). Thank you!
